pipeline-runner/R/gem2s-2-load_user_files.R

```diff
@@ -234,3 +254,84 @@ format_annot <- function(annot_list) {
+#  0 is SYMBOL/ SYMBOL
+#  1 is IDS/SYMBOL
+#  2 is IDS/IDS
+equalize_annotation_types <- function(annot_list, counts_list, features_types_list, samples) {
+  if (any(features_types_list == 2) && any(features_types_list == 0) && !any(features_types_list == 1)) stop("Incompatible features detected.")
+
+  if (any(features_types_list == 1) && (any(features_types_list == 2) || any(features_types_list == 0))) {
+    annots_with_ids <- unique(do.call("rbind", annot_list[features_types_list == 1]))
+
+    annots_with_ids <- annots_with_ids[!duplicated(annots_with_ids$input), ]
+
+    for (sample in samples) {
+      if (features_types_list[[sample]] == 0 || features_types_list[[sample]] == 2) {
+        sample_annot <- annot_list[[sample]]
+
+        # Try to replace input column (currently symbols) in sample_annot with ids from annots_with_ids
+        if (features_types_list[[sample]] == 0) {
+
+          matched_symbols_index <- match(sample_annot$input,annots_with_ids$name)
+          is_in_annot_list <- which(sample_annot$input %in% annots_with_ids$name)
+
+          sample_annot$input[is_in_annot_list] <- annots_with_ids$input[na.omit(matched_symbols_index)]
+
+          #This avoids duplicates after combining with the annotated df.
+          #Leads to a mismatch in genes between samples but it seems like the best solution
+          sample_annot$input <- make.unique(sample_annot$input)
+
+          #The counts have been loaded with gene symbols so we need to replace the rownames with the ids
+          rownames(counts_list[[sample]]) <- sample_annot$input
+        }
+
+        # Try to replace names column (currently ids) in sample_annot with symbols from annots_with_ids
+        if (features_types_list[[sample]] == 2) {
+          matched_ids_index <- match(sample_annot$input,annots_with_ids$input)
+          is_in_annot_list <- which(sample_annot$input %in% annots_with_ids$input)
+
+          sample_annot$name[is_in_annot_list] <- annots_with_ids$name[na.omit(matched_ids_index)]
+        }
+
+        annot_list[[sample]] <- sample_annot
+      }
+    }
+  }
+  return(list(counts_list=counts_list, annot_list=annot_list))
+}
+
+#' extract_features_types
+#'
+#' Determines the type of an annot data frame.
+#' Classifies the columns into either ensemblIds or symbols, and extracts a number that represents the combination.
+#'
+#' @param annot
+#'
+#' @return
+#' -1 is SYMBOl/IDS
+#' 0 is SYMBOL/ SYMBOL
+#' 1 is IDS/SYMBOL
+#' 2 is IDS/IDS
+#'
+#'
+#' @examples
+extract_features_types <- function(annot) {
+  features_types <- list()
+
+  annot_c1 <- annot[, 1]
+  is_ens <- annot_c1[substr(annot_c1, 1, 3) == "ENS"]
+  features_types[[1]] <- length(is_ens) >= length(annot_c1) - length(is_ens)
+
+  annot_c2 <- annot[, 2]
+  is_ens <- annot_c2[substr(annot_c2, 1, 3) == "ENS"]
+  features_types[[2]] <- length(is_ens) >= length(annot_c2) - length(is_ens)
+
+  if (features_types[[1]] == FALSE && features_types[[2]] == TRUE) {
```

Comment:
more than half of values in column start with string ENS?

```R
is_ens_col1 <- startsWith(annot[[1]], "ENS")
pct_ens_col1 <- sum(is_ens_col1) / nrow(annot)

is_ens_col2 <- startsWith(annot[[2]], "ENS")
pct_ens_col2 <- sum(is_ens_col2) / nrow(annot)

if (pct_ens_col1 > 0.5 & pct_ens_col2 > 0.5) return(IDS_IDS)
```

Reply:
I like this. stolen 👍